### ISSUE TITLE
[1.3.0-RC4] StackOverflowError in trigerred tasks

### ISSUE THREAD

**Issue by OlegYch** [sbt/sbt#4970]
## steps
sbt version: 1.3.0-RC4

## problem
running any task with `~` produces 
```
[debug] > Exec(~update, None, None)
[debug] Evaluating tasks: checkBuildSources
[debug] Running task... Cancel: Signal, check cycles: false, forcegc: true
[debug] Forcing garbage collection...
[debug] Running task... Cancel: Signal, check cycles: false, forcegc: true
[error] java.lang.StackOverflowError
[error] 	at sbt.nio.file.Glob$ordering$.compare(Glob.scala:270)
[error] 	at sbt.nio.file.Glob$ordering$.compare(Glob.scala:265)
[error] 	at sbt.nio.file.Glob$ordering$.compare(Glob.scala:270)
[error] 	at sbt.nio.file.Glob$ordering$.compare(Glob.scala:265)
[error] 	at sbt.nio.file.Glob$ordering$.compare(Glob.scala:270)

```

## expectation
no error

## notes
this is a play project using play-sbt-plugin 2.7.3
this was working with RC3

**Comment by OlegYch** [comment]
here is a repro project https://github.com/OlegYch/sbt-4970

**Comment by eatkins** [comment]
Thanks for the repro. It's an easy fix.

**Comment by OlegYch** [comment]
doesn't look like this is fully fixed
```
[debug] > Exec(~compile, None, None)
[debug] Evaluating tasks: checkBuildSources
[debug] Running task... Cancel: Signal, check cycles: false, forcegc: true
[debug] Running task... Cancel: Signal, check cycles: false, forcegc: true
[error] java.lang.IllegalArgumentException: Comparison method violates its general contract!
[error] 	at java.util.TimSort.mergeHi(TimSort.java:899)
[error] 	at java.util.TimSort.mergeAt(TimSort.java:516)
[error] 	at java.util.TimSort.mergeForceCollapse(TimSort.java:457)
[error] 	at java.util.TimSort.sort(TimSort.java:254)
[error] 	at java.util.Arrays.sort(Arrays.java:1438)
[error] 	at scala.collection.SeqLike.sorted(SeqLike.scala:659)
[error] 	at scala.collection.SeqLike.sorted$(SeqLike.scala:647)
[error] 	at scala.collection.AbstractSeq.sorted(Seq.scala:45)
[error] 	at sbt.internal.SettingsGraph$.transitiveDynamicInputs(SettingsGraph.scala:128)
[error] 	at sbt.internal.Continuous$.getConfig(Continuous.scala:180)
[error] 	at sbt.internal.Continuous$.$anonfun$getAllConfigs$3(Continuous.scala:423)
[error] 	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:237)
[error] 	at scala.collection.immutable.List.foreach(List.scala:392)
[error] 	at scala.collection.TraversableLike.map(TraversableLike.scala:237)
[error] 	at scala.collection.TraversableLike.map$(TraversableLike.scala:230)
[error] 	at scala.collection.immutable.List.map(List.scala:298)
[error] 	at sbt.internal.Continuous$.$anonfun$getAllConfigs$2(Continuous.scala:423)
[error] 	at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:244)
[error] 	at scala.collection.immutable.List.foreach(List.scala:392)
[error] 	at scala.collection.TraversableLike.flatMap(TraversableLike.scala:244)
[error] 	at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:241)
[error] 	at scala.collection.immutable.List.flatMap(List.scala:355)
[error] 	at sbt.internal.Continuous$.getAllConfigs(Continuous.scala:422)
[error] 	at sbt.internal.Continuous$.$anonfun$runToTermination$7(Continuous.scala:347)
[error] 	at sbt.std.Streams.use(Streams.scala:89)
[error] 	at sbt.std.Streams.use$(Streams.scala:85)
[error] 	at sbt.std.Streams$$anon$1.use(Streams.scala:102)
[error] 	at sbt.internal.Continuous$.$anonfun$runToTermination$6(Continuous.scala:343)
[error] 	at sbt.EvaluateTask$.withStreams(EvaluateTask.scala:373)
[error] 	at sbt.internal.Continuous$.$anonfun$runToTermination$5(Continuous.scala:343)
[error] 	at sbt.internal.Continuous$.setup(Continuous.scala:273)
[error] 	at sbt.internal.Continuous$.$anonfun$runToTermination$1(Continuous.scala:342)
[error] 	at sbt.internal.Continuous$.withCharBufferedStdIn(Continuous.scala:307)
[error] 	at sbt.internal.Continuous$.runToTermination(Continuous.scala:321)
[error] 	at sbt.internal.Continuous$.$anonfun$continuous$1(Continuous.scala:108)
[error] 	at sbt.Command$.$anonfun$applyEffect$4(Command.scala:149)
[error] 	at sbt.Command$.$anonfun$applyEffect$2(Command.scala:144)
[error] 	at sbt.Command$.process(Command.scala:187)
[error] 	at sbt.MainLoop$.process$1(MainLoop.scala:185)
[error] 	at sbt.MainLoop$.processCommand(MainLoop.scala:218)
[error] 	at sbt.MainLoop$.$anonfun$next$2(MainLoop.scala:142)
[error] 	at sbt.State$StateOpsImpl$.runCmd$1(State.scala:273)
[error] 	at sbt.State$StateOpsImpl$.process$extension(State.scala:277)
[error] 	at sbt.MainLoop$.$anonfun$next$1(MainLoop.scala:142)
[error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:19)
[error] 	at sbt.MainLoop$.next(MainLoop.scala:142)
[error] 	at sbt.MainLoop$.run(MainLoop.scala:133)
[error] 	at sbt.MainLoop$.$anonfun$runWithNewLog$1(MainLoop.scala:111)
[error] 	at sbt.io.Using.apply(Using.scala:27)
[error] 	at sbt.MainLoop$.runWithNewLog(MainLoop.scala:105)
[error] 	at sbt.MainLoop$.runAndClearLast(MainLoop.scala:60)
[error] 	at sbt.MainLoop$.runLoggedLoop(MainLoop.scala:45)
[error] 	at sbt.MainLoop$.runLogged(MainLoop.scala:36)
[error] 	at sbt.StandardMain$.runManaged(Main.scala:132)
[error] 	at sbt.xMain$.run(Main.scala:67)
[error] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[error] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[error] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[error] 	at java.lang.reflect.Method.invoke(Method.java:498)
[error] 	at sbt.internal.XMainConfiguration.run(XMainConfiguration.scala:46)
[error] 	at sbt.xMain.run(Main.scala:39)
[error] 	at xsbt.boot.Launch$$anonfun$run$1.apply(Launch.scala:109)
[error] 	at xsbt.boot.Launch$.withContextLoader(Launch.scala:128)
[error] 	at xsbt.boot.Launch$.run(Launch.scala:109)
[error] 	at xsbt.boot.Launch$$anonfun$apply$1.apply(Launch.scala:35)
[error] 	at xsbt.boot.Launch$.launch(Launch.scala:117)
[error] 	at xsbt.boot.Launch$.apply(Launch.scala:18)
[error] 	at xsbt.boot.Boot$.runImpl(Boot.scala:56)
[error] 	at xsbt.boot.Boot$.main(Boot.scala:18)
[error] 	at xsbt.boot.Boot.main(Boot.scala)
[error] java.lang.IllegalArgumentException: Comparison method violates its general contract!
[error] Use 'last' for the full log.
[debug] > Exec(idea-shell, None, None)
[debug] Evaluating tasks: checkBuildSources
[debug] Running task... Cancel: Signal, check cycles: false, forcegc: true
```

**Comment by OlegYch** [comment]
removing
```
    sourceDirectories in (Compile, TwirlKeys.compileTemplates) := Seq(baseDirectory.value / "templates"),
    unmanagedSourceDirectories in Compile += baseDirectory.value / "templates",
```
from my project seems to fix `~compile` (fwiw templates is an empty dir anyway), but i can't seem to create a minimal repro since removing sources also fixes it

**Comment by OlegYch** [comment]
`~test:compile` still fails though

**Comment by OlegYch** [comment]
on another machine `~test:compile` works but `~compile` fails...

**Comment by eatkins** [comment]
@OlegYch #5012 should fix the issue, but just out of curiosity does it go away if you set `excludeFilter in TwirlKeys.compileTemplates := NothingFilter`?.

**Comment by OlegYch** [comment]
@eatkins nope
btw i'm pretty sure DynamicInput ordering is correct up to Glob ordering, so if Glob ordering remains the same the issue might pop up somewhere else

**Comment by eatkins** [comment]
That is accurate. I will deprecate glob ordering in io, but within sbt it was only used by DynamicInput ordering.

**Comment by eatkins** [comment]
@OlegYch Any chance you can try publishing a local version with this change and report back? I can't actually reproduce.

**Comment by eatkins** [comment]
Actually, I think I understand what was causing the issue: https://github.com/sbt/io/pull/267.

**Comment by eatkins** [comment]
@OlegYch Would it be possible for you to share the result of `show watchSources` in your project? I'm trying to write a scripted test and I'm finding it difficult to reproduce your crashes even without https://github.com/sbt/io/pull/267.

**Comment by eatkins** [comment]
You can leave out individual files, e.g. 
```
Source(
  base = /private/var/folders/7q/s3vq91lx3zv9fw4l5l5c4t_f6a97cc2/src,
  includeFilter = ExactFilter(Foo.scala),
  excludeFilter = NothingFilter,
  recursive = false,
)
```
I'm trying to figure out specifically what conversions from `Source` -> `Glob` cause problems in your build.

**Comment by OlegYch** [comment]
@eatkins i'll try, but it sounds like scalachecking glob ordering could do it faster

**Comment by eatkins** [comment]
Maybe. Except that I've never used scalacheck so I'd have to learn a new tool.

**Comment by OlegYch** [comment]
@eatkins i pushed a sample of 'watchSources' to my fork, alongside an attempt of scalacheck test
i'm afraid it isn't of much use since toString is not really informative
wonder why every source file has an entry there though, instead of one app/**.scala glob

**Comment by eatkins** [comment]
Cool. That's very helpful. Thank you!

> wonder why every source file has an entry there though, instead of one app/**.scala glob

It's because of the default layout of a play project. They unfortunately put the assets directory under the main source directory so they had to filter the assets out:
```
watchSources ++= {
  ((sourceDirectory in Compile).value ** "*" --- (sourceDirectory in Assets).value ** "*").get
},
```
That being said, they could have done:
```
watchSources += { 
  val assetsDir = (sourceDirectory in Assets).value.toPath
  WatchSource(
    (sourceDirectory in Compile).value,
    includeFilter = AllPassFilter,
    excludeFilter = new SimpleFileFilter(file => file.toPath.startsWith(assetsDir))
    recursive = true
  )
}
```
I probably should open a PR over there.

**Comment by eatkins** [comment]
@OlegYch is this fixed in 1.3.0?

**Comment by OlegYch** [comment]
Can't do much testing right now, but it seems so. I think you can close.On Sep 5, 2019 22:15, Ethan Atkins <notifications@github.com> wrote:@OlegYch is this fixed in 1.3.0?

**Comment by eatkins** [comment]
Ok. Closing for now. Feel free to re-open if it crops up again.
